cp.x: implemented new order of atoms in trajectory, AUTOPILOT input, new XML output

rikigigi opened the pull request:
Autopilot was missing in the cp input. Is this tentative implementation ok? I added in the PW/CP input generator a call to a function that is implemented in the derived class.

sample parameters:
```
{
    'AUTOPILOT' : [
        {'onstep' : 10, 'what' : 'dt', 'newvalue' : 5.0 },
        {'onstep' : 20, 'what' : 'dt', 'newvalue' : 10.0}
    ],
}
```

rikigigi commented:
Should pw and cp use the same XML parser? Work in progress.

rikigigi commented:
Many of these modifications are **needed** to use cp.x in the last qe version.

I implemented changed that are needed in order to use aiida with cp in the last version.
Note that the order of the atoms in the cp trajectory output changed somewhere after qe version 6.5. Since the version in the qe output is still 6.5, there is no way to check which kind of trajectory is produced with the git version, so right now it may not work with the qe git version if in the input the atoms are not ordered by type.
Then there is the possibility that the trajectory output is not present, or the print_counter file is not present. This means simply that no trajectory was written to the filesystem, but the program maybe did something (like a single cg step). So in this version the parser does not fail if it does not find the trajectory output.
I implemented reading of the new version of the xml file using pw parser function.
Then now it is present the input for the autopilot card in cp, that can decrease the number of different inputs that need to be generated and submitted.
The reading of the forces from the trajectory is implemented too.

Now I think that is finished.

giovannipizzi commented:
Hi @rikigigi, sorry for the time it's taking to review this! This was caught in between a number of releases etc.

Is it ok to maybe have a short call to discuss the changes? I've not been using CP in the past couple of years, and wanted to check with you the changes, to understand briefly backward-compatibility for earlier versions etc.
You can drop me a line at my email giovanni.pizzi@epfl.ch if you are interested

Pinging also @sphuber in case he wants to participate (and @mkotiuga as she might be interested in running with the cp.x plugin as well)

lorisercole commented:
Hey @rikigigi, just a quick question (I haven't looked at your edits yet).
Have you considered the fact that the `.cel` trajectory files of CP (for whatever reason) contain _transposed_ cell matrices? (see the [issue in QE](https://gitlab.com/QEF/q-e/-/issues/132)).
I think once I noticed that the CP parser was not considering this oddness, but I never took the time to amend the code and verify it. 
Thanks! ;)

rikigigi commented:
Yes, this was missing in the previous code

Riccardo Bertossa

Il gio 28 mag 2020, 23:02 Loris Ercole <notifications@github.com> ha
scritto:

> Hey @rikigigi <https://github.com/rikigigi>, just a quick question (I
> haven't looked at your edits yet).
> Have you considered the fact that the .cel trajectory files of CP (for
> whatever reason) contain *transposed* cell matrices? (see the issue in QE
> <https://gitlab.com/QEF/q-e/-/issues/132>).
> I think once I noticed that the CP parser was not considering this
> oddness, but I never took the time to amend the code and verify it.
> Thanks! ;)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aiidateam/aiida-quantumespresso/pull/455#issuecomment-635605990>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIBKSWMADVLABQ3RB2GLGCDRT3GOXANCNFSM4KIGGHMA>
> .
>

giovannipizzi commented on aiida_quantumespresso/parsers/cp.py:
```diff
@@ -5,10 +5,11 @@
 from aiida.orm import Dict, TrajectoryData
 
-from qe_tools.constants import bohr_to_ang, hartree_to_ev, timeau_to_sec
+from qe_tools.constants import bohr_to_ang, hartree_to_ev, timeau_to_sec, hartree_to_ev
```
double import

giovannipizzi commented on aiida_quantumespresso/parsers/cp.py:
```diff
@@ -39,140 +40,181 @@ def parse(self, **kwargs):
+
+        print_counter_xml=True
+        no_trajectory_output=False
+        if self.node.process_class._FILE_XML_PRINT_COUNTER_BASENAME not in list_of_files and self.node.process_class._FILE_PRINT_COUNTER_BASENAME not in list_of_files:
```
Maybe documentation (see above) could be here

giovannipizzi commented on aiida_quantumespresso/parsers/cp.py:
```diff
@@ -39,140 +40,181 @@ def parse(self, **kwargs):
+            self.logger.error('We could not find the print counter file in the output (' + self.node.process_class._FILE_PRINT_COUNTER_BASENAME + ' or ' + self.node.process_class._FILE_XML_PRINT_COUNTER_BASENAME + ' not in {} )'.format(list_of_files) + 'assuming no trajectory output was produced' )
+            no_trajectory_output=True
+            #this can happen and it is not an error!
+        if not no_trajectory_output:
```
Maybe at least reference one version in which you know the given format was used

giovannipizzi commented on aiida_quantumespresso/parsers/cp.py:
```diff
@@ -39,140 +40,181 @@ def parse(self, **kwargs):
+
+            #order of atom in the output trajectory changed somewhere after 6.5
+            if LooseVersion(out_dict['creator_version']) > LooseVersion('6.5'):
+                new_cp_ordering = True
```
Let's see if we can push cp/QE to have a variable in the XML output (or as a head of each file) to specify the format

rikigigi replied:
I think we are going to use 6.5 as the last versione where we have the reordering of the atoms, with no new variables in the XML output

giovannipizzi commented on aiida_quantumespresso/parsers/cp.py:
```diff
@@ -39,140 +40,181 @@ def parse(self, **kwargs):
+                            reordering
+                        )
+                    else:
+                        raw_trajectory['cells'] = numpy.array(traj_data['cells_traj_data']).transpose((0,2,1))
```
Add a comment

giovannipizzi commented on aiida_quantumespresso/parsers/cp.py:
```diff
@@ -39,140 +40,181 @@ def parse(self, **kwargs):
-                # Some columns may have not been parsed, skip
-                pass
+                traj.set_array('forces', raw_trajectory['forces_ordered'])
+            except:
```
Let's use a more specific exception (KeyError?). Moreover, could you check what happens if the file is empty?

giovannipizzi commented on aiida_quantumespresso/parsers/parse_raw/cp.py:
```diff
@@ -81,54 +88,56 @@ def parse_cp_text_output(data, xml_data):
-                parsed_data['xnhp0'] = [float(this_line.split()[11])]
-            except (ValueError, IndexError):
-                pass
+    try:
```
Is this still needed, and when? If you have a reproducible case, maybe we should see if we can skip according to a flag, or put some message in the parsed output to say we skipped for reason XXX

rikigigi replied:
The parser works only for the verlet output. When the code is doing cg steps, the output is different, and maybe we need a new parser for that

giovannipizzi commented on aiida_quantumespresso/parsers/parse_raw/cp.py:
```diff
@@ -148,21 +157,43 @@ def parse_cp_xml_counter_output(data):
     return parsed_data
 
+def parse_cp_counter_output(data):
+    """Parse xml file print_counter data must be a single string, as returned by file.read() (notice the difference
```
Fix comment (no XML).

giovannipizzi commented on aiida_quantumespresso/parsers/parse_raw/cp.py:
```diff
@@ -204,112 +235,176 @@ def parse_cp_xml_output(data):
 
     #CARD HEADER
-    parsed_data = copy.deepcopy(xml_card_header(parsed_data, dom))
+    try:
```
I suggest to remove all the try/except blocks, move the parsing of the old XML (legacy) in a file raw_cp_legacy, and call it only depending on the XML version.

rikigigi replied:
I found that they were just harmless wrecks of some ancient attempt to write a happy parser. Removed

giovannipizzi commented on aiida_quantumespresso/parsers/parse_raw/cp.py:
```diff
@@ -204,112 +235,176 @@ def parse_cp_xml_output(data):
     attrname = 'ITERATION'
-    parsed_data[(tagname + '_' + attrname).lower()] = int(parse_xml_child_attribute_str(tagname, attrname, target_tags))
+    try:
+        parsed_data[(tagname+'_'+attrname).lower()]=int(parse_xml_child_attribute_str(tagname,attrname,target_tags))
```
Is iteration a CP specific parameter, and do you know how it's called now?

giovannipizzi commented on aiida_quantumespresso/parsers/parse_raw/cp.py:
```diff
@@ -204,112 +235,176 @@ def parse_cp_xml_output(data):
-    if units not in ['pico-seconds']:
-        raise QEOutputParsingError('Units {} are not supported by parser'.format(units))
-    parsed_data[tagname.lower()] = value
+    try:
```
Is this already parsed by the new XML parser?

rikigigi replied:
yes

giovannipizzi commented on aiida_quantumespresso/parsers/parse_raw/cp.py:
```diff
@@ -347,7 +442,11 @@ def parse_cp_xml_output(data):
 
     cardname = 'TIMESTEPS'
-    target_tags = read_xml_card(dom, cardname)
+    try:
```
Is this parsed in the new XML?

giovannipizzi commented on aiida_quantumespresso/parsers/parse_xml/pw/parse.py:
```diff
@@ -101,7 +101,7 @@ def parse_pw_xml_post_6_2(xml):
 
     xml_version = StrictVersion(xml_dictionary['general_info']['xml_format']['@VERSION'])
-    inputs = xml_dictionary['input']
+    inputs = xml_dictionary.get('input',{})
```
Can we check with QE (Pietro Delugas?) if these changes are intentional?

rikigigi replied:
The xml files that I have do not have the input section

giovannipizzi commented on tests/parsers/test_cp/test_cp_default.yml:
```diff
@@ -208,6 +208,8 @@ trajectory:
   array|enthalpy_plus_kinetic:
   - 10
+  array|forces:
+  - 0
```
Check if we want not to store the forces if the file is empty

rikigigi commented on tests/parsers/test_cp.py:
```diff
@@ -7,32 +7,46 @@
+
+#TODO: update the output when the version after 6.5 is out! Now I used a post-6.5 git version (order of atom in the output changed)
+@pytest.mark.parametrize('version',['default','6.5_autopilot','6.5','6.5_cgstep','6.5_cgsteps'])
 def test_cp_default(
```
I extended this test. The old one is `test_cp_default` with `version='default`, and it is working.
Is someone more experienced in pytest than me able to tell why two of the others (autopilot and cgstepS) fails with `yaml.representer.RepresenterError: ('cannot represent an object', array([[[ --bands here-- ]]])`? Full output of the test below (the other fails are due to other issues that I have to address)
```
 (aiida-venv) bertossa@pippo:~/sissa/aiida-quantumespresso$ pytest -sv tests/ -k 'test_cp_default'
========================================================================== test session starts ===========================================================================
platform linux -- Python 3.6.9, pytest-5.4.2, py-1.8.1, pluggy-0.13.1 -- /home/bertossa/sissa/aiida-venv/bin/python3
cachedir: .pytest_cache
rootdir: /home/bertossa/sissa/aiida-quantumespresso/tests, inifile: pytest.ini
plugins: regressions-1.0.6, datadir-1.3.1
collected 130 items / 125 deselected / 5 selected                                                                                                                        

tests/parsers/test_cp.py::test_cp_default[default] PASSED
tests/parsers/test_cp.py::test_cp_default[6.5_autopilot] list index out of range
FAILED
tests/parsers/test_cp.py::test_cp_default[6.5] FAILED
tests/parsers/test_cp.py::test_cp_default[6.5_cgstep] list index out of range
FAILED
tests/parsers/test_cp.py::test_cp_default[6.5_cgsteps] list index out of range
FAILED

================================================================================ FAILURES ================================================================================
_____________________________________________________________________ test_cp_default[6.5_autopilot] _____________________________________________________________________

aiida_profile = <aiida.manage.tests.TestManager object at 0x7fa234702a90>, fixture_localhost = <Computer: localhost-test (localhost-test), pk: 1>
generate_calc_job_node = <function generate_calc_job_node.<locals>._generate_calc_job_node at 0x7fa22097d730>
generate_parser = <function generate_parser.<locals>._generate_parser at 0x7fa21ffc7510>
data_regression = <pytest_regressions.data_regression.DataRegressionFixture object at 0x7fa22044d0f0>
generate_structure = <function generate_structure.<locals>._generate_structure at 0x7fa2203f2730>, version = '6.5_autopilot'

    @pytest.mark.parametrize('version',['default','6.5_autopilot','6.5','6.5_cgstep','6.5_cgsteps'])
    def test_cp_default(
        aiida_profile, fixture_localhost, generate_calc_job_node, generate_parser, data_regression, generate_structure, version
    ):
        """Test a default `cp.x` calculation."""
        entry_point_calc_job = 'quantumespresso.cp'
        entry_point_parser = 'quantumespresso.cp'
        if version=='default':
            def generate_inputs():
                return AttributeDict({
                    'structure': generate_structure(structure_id='silicon'),
                    'parameters': orm.Dict(dict={}),
                })
        else:
            def generate_inputs():
                return AttributeDict({
                    'structure': generate_structure(structure_id='water'),
                    'parameters': orm.Dict(dict={}),
                })
    
        node = generate_calc_job_node(entry_point_calc_job, fixture_localhost, version, generate_inputs())
        parser = generate_parser(entry_point_parser)
        results, calcfunction = parser.parse_from_node(node, store_provenance=False)
    
        assert calcfunction.is_finished, calcfunction.exception
        assert calcfunction.is_finished_ok, calcfunction.exit_message
        assert not orm.Log.objects.get_logs_for(node)
        assert 'output_parameters' in results
        if version is not 'cgstep':
            assert 'output_trajectory' in results
            data_regression.check({
                'parameters': results['output_parameters'].get_dict(),
>               'trajectory': results['output_trajectory'].attributes
            })

tests/parsers/test_cp.py:46: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../aiida-venv/lib/python3.6/site-packages/pytest_regressions/data_regression.py:55: in dump
    encoding="utf-8",
../aiida-venv/lib/python3.6/site-packages/yaml/__init__.py:278: in dump_all
    dumper.represent(data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:27: in represent
    node = self.represent_data(data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:48: in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:207: in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:118: in represent_mapping
    node_value = self.represent_data(item_value)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:48: in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:207: in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:118: in represent_mapping
    node_value = self.represent_data(item_value)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:48: in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:207: in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:118: in represent_mapping
    node_value = self.represent_data(item_value)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:58: in represent_data
    node = self.yaml_representers[None](self, data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pytest_regressions.data_regression.RegressionYamlDumper object at 0x7fa220aae710>, data = array([[[-12.07535278,  -6.28316664,  -4.02636205,  -3.17507374]]])

    def represent_undefined(self, data):
>       raise RepresenterError("cannot represent an object", data)
E       yaml.representer.RepresenterError: ('cannot represent an object', array([[[-12.07535278,  -6.28316664,  -4.02636205,  -3.17507374]]]))

../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:231: RepresenterError
__________________________________________________________________________ test_cp_default[6.5] __________________________________________________________________________

aiida_profile = <aiida.manage.tests.TestManager object at 0x7fa234702a90>, fixture_localhost = <Computer: localhost-test (localhost-test), pk: 1>
generate_calc_job_node = <function generate_calc_job_node.<locals>._generate_calc_job_node at 0x7fa22035c8c8>
generate_parser = <function generate_parser.<locals>._generate_parser at 0x7fa21ffc7510>
data_regression = <pytest_regressions.data_regression.DataRegressionFixture object at 0x7fa220aae1d0>
generate_structure = <function generate_structure.<locals>._generate_structure at 0x7fa22035c1e0>, version = '6.5'

    @pytest.mark.parametrize('version',['default','6.5_autopilot','6.5','6.5_cgstep','6.5_cgsteps'])
    def test_cp_default(
        aiida_profile, fixture_localhost, generate_calc_job_node, generate_parser, data_regression, generate_structure, version
    ):
        """Test a default `cp.x` calculation."""
        entry_point_calc_job = 'quantumespresso.cp'
        entry_point_parser = 'quantumespresso.cp'
        if version=='default':
            def generate_inputs():
                return AttributeDict({
                    'structure': generate_structure(structure_id='silicon'),
                    'parameters': orm.Dict(dict={}),
                })
        else:
            def generate_inputs():
                return AttributeDict({
                    'structure': generate_structure(structure_id='water'),
                    'parameters': orm.Dict(dict={}),
                })
    
        node = generate_calc_job_node(entry_point_calc_job, fixture_localhost, version, generate_inputs())
        parser = generate_parser(entry_point_parser)
        results, calcfunction = parser.parse_from_node(node, store_provenance=False)
    
        assert calcfunction.is_finished, calcfunction.exception
>       assert calcfunction.is_finished_ok, calcfunction.exit_message
E       AssertionError: The stdout output file could not be read.
E       assert False
E        +  where False = <CalcFunctionNode: uuid: 474ab038-9f09-474a-be4b-86bd7493a225 (unstored) (aiida.parsers.parser.parse_calcfunction)>.is_finished_ok

tests/parsers/test_cp.py:39: AssertionError
______________________________________________________________________ test_cp_default[6.5_cgstep] _______________________________________________________________________

aiida_profile = <aiida.manage.tests.TestManager object at 0x7fa234702a90>, fixture_localhost = <Computer: localhost-test (localhost-test), pk: 1>
generate_calc_job_node = <function generate_calc_job_node.<locals>._generate_calc_job_node at 0x7fa21fb94d90>
generate_parser = <function generate_parser.<locals>._generate_parser at 0x7fa21ffc7510>
data_regression = <pytest_regressions.data_regression.DataRegressionFixture object at 0x7fa220a4aeb8>
generate_structure = <function generate_structure.<locals>._generate_structure at 0x7fa21fb94ea0>, version = '6.5_cgstep'

    @pytest.mark.parametrize('version',['default','6.5_autopilot','6.5','6.5_cgstep','6.5_cgsteps'])
    def test_cp_default(
        aiida_profile, fixture_localhost, generate_calc_job_node, generate_parser, data_regression, generate_structure, version
    ):
        """Test a default `cp.x` calculation."""
        entry_point_calc_job = 'quantumespresso.cp'
        entry_point_parser = 'quantumespresso.cp'
        if version=='default':
            def generate_inputs():
                return AttributeDict({
                    'structure': generate_structure(structure_id='silicon'),
                    'parameters': orm.Dict(dict={}),
                })
        else:
            def generate_inputs():
                return AttributeDict({
                    'structure': generate_structure(structure_id='water'),
                    'parameters': orm.Dict(dict={}),
                })
    
        node = generate_calc_job_node(entry_point_calc_job, fixture_localhost, version, generate_inputs())
        parser = generate_parser(entry_point_parser)
        results, calcfunction = parser.parse_from_node(node, store_provenance=False)
    
        assert calcfunction.is_finished, calcfunction.exception
        assert calcfunction.is_finished_ok, calcfunction.exit_message
>       assert not orm.Log.objects.get_logs_for(node)
E       AssertionError: assert not [<aiida.orm.logs.Log object at 0x7fa21faf5160>]
E        +  where [<aiida.orm.logs.Log object at 0x7fa21faf5160>] = <bound method Log.Collection.get_logs_for of <aiida.orm.logs.Log.Collection object at 0x7fa22073b358>>(<CalcJobNode: uuid: 44058aef-73e1-4097-a73c-986166f9c16e (pk: 18) (aiida.calculations:quantumespresso.cp)>)
E        +    where <bound method Log.Collection.get_logs_for of <aiida.orm.logs.Log.Collection object at 0x7fa22073b358>> = <aiida.orm.logs.Log.Collection object at 0x7fa22073b358>.get_logs_for
E        +      where <aiida.orm.logs.Log.Collection object at 0x7fa22073b358> = <class 'aiida.orm.logs.Log'>.objects
E        +        where <class 'aiida.orm.logs.Log'> = orm.Log

tests/parsers/test_cp.py:40: AssertionError
______________________________________________________________________ test_cp_default[6.5_cgsteps] ______________________________________________________________________

aiida_profile = <aiida.manage.tests.TestManager object at 0x7fa234702a90>, fixture_localhost = <Computer: localhost-test (localhost-test), pk: 1>
generate_calc_job_node = <function generate_calc_job_node.<locals>._generate_calc_job_node at 0x7fa21fa7c950>
generate_parser = <function generate_parser.<locals>._generate_parser at 0x7fa21ffc7510>
data_regression = <pytest_regressions.data_regression.DataRegressionFixture object at 0x7fa21fb45eb8>
generate_structure = <function generate_structure.<locals>._generate_structure at 0x7fa21fa7cd90>, version = '6.5_cgsteps'

    @pytest.mark.parametrize('version',['default','6.5_autopilot','6.5','6.5_cgstep','6.5_cgsteps'])
    def test_cp_default(
        aiida_profile, fixture_localhost, generate_calc_job_node, generate_parser, data_regression, generate_structure, version
    ):
        """Test a default `cp.x` calculation."""
        entry_point_calc_job = 'quantumespresso.cp'
        entry_point_parser = 'quantumespresso.cp'
        if version=='default':
            def generate_inputs():
                return AttributeDict({
                    'structure': generate_structure(structure_id='silicon'),
                    'parameters': orm.Dict(dict={}),
                })
        else:
            def generate_inputs():
                return AttributeDict({
                    'structure': generate_structure(structure_id='water'),
                    'parameters': orm.Dict(dict={}),
                })
    
        node = generate_calc_job_node(entry_point_calc_job, fixture_localhost, version, generate_inputs())
        parser = generate_parser(entry_point_parser)
        results, calcfunction = parser.parse_from_node(node, store_provenance=False)
    
        assert calcfunction.is_finished, calcfunction.exception
        assert calcfunction.is_finished_ok, calcfunction.exit_message
        assert not orm.Log.objects.get_logs_for(node)
        assert 'output_parameters' in results
        if version is not 'cgstep':
            assert 'output_trajectory' in results
            data_regression.check({
                'parameters': results['output_parameters'].get_dict(),
>               'trajectory': results['output_trajectory'].attributes
            })

tests/parsers/test_cp.py:46: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../aiida-venv/lib/python3.6/site-packages/pytest_regressions/data_regression.py:55: in dump
    encoding="utf-8",
../aiida-venv/lib/python3.6/site-packages/yaml/__init__.py:278: in dump_all
    dumper.represent(data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:27: in represent
    node = self.represent_data(data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:48: in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:207: in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:118: in represent_mapping
    node_value = self.represent_data(item_value)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:48: in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:207: in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:118: in represent_mapping
    node_value = self.represent_data(item_value)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:48: in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:207: in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:118: in represent_mapping
    node_value = self.represent_data(item_value)
../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:58: in represent_data
    node = self.yaml_representers[None](self, data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pytest_regressions.data_regression.RegressionYamlDumper object at 0x7fa21fa2cda0>, data = array([[[0., 0., 0., 0.]]])

    def represent_undefined(self, data):
>       raise RepresenterError("cannot represent an object", data)
E       yaml.representer.RepresenterError: ('cannot represent an object', array([[[0., 0., 0., 0.]]]))

../aiida-venv/lib/python3.6/site-packages/yaml/representer.py:231: RepresenterError
============================================================================ warnings summary ============================================================================
/home/bertossa/sissa/aiida-venv/lib/python3.6/site-packages/past/builtins/misc.py:45
  /home/bertossa/sissa/aiida-venv/lib/python3.6/site-packages/past/builtins/misc.py:45: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    from imp import reload

-- Docs: https://docs.pytest.org/en/latest/warnings.html
======================================================================== short test summary info =========================================================================
FAILED tests/parsers/test_cp.py::test_cp_default[6.5_autopilot] - yaml.representer.RepresenterError: ('cannot represent an object', array([[[-12.07535278,  -6.28316664...
FAILED tests/parsers/test_cp.py::test_cp_default[6.5] - AssertionError: The stdout output file could not be read.
FAILED tests/parsers/test_cp.py::test_cp_default[6.5_cgstep] - AssertionError: assert not [<aiida.orm.logs.Log object at 0x7fa21faf5160>]
FAILED tests/parsers/test_cp.py::test_cp_default[6.5_cgsteps] - yaml.representer.RepresenterError: ('cannot represent an object', array([[[0., 0., 0., 0.]]]))
========================================================= 4 failed, 1 passed, 125 deselected, 1 warning in 7.74s =========================================================
(aiida-venv) bertossa@pippo:~/sissa/aiida-quantumespresso$ 

```

@giovannipizzi 
@sphuber 
@ltalirz 
@lorisercole

giovannipizzi replied:
Only native python types can be yaml-serialized, not numpy arrays. Probably there is a way to pass arrays in some way, but the simplest solution is to call the `.tolist()` method of the numpy arrays, so they are converted to lists (or lists of lists if they are 2D), and can be serialised. You are then verifying that the values of the list of lists are the same, but this has the same effect.

rikigigi replied:
Ok, I'll try something like this. But were the previous parser's tests parsing any numpy array?

sphuber replied:
The weird thing is that it excepts at `'trajectory': results['output_trajectory'].attributes`, but the attributes of a `TrajectoryData` node *should* be serializable. We do the exact same thing for the other tests. So the attributes must contain something weird, but cannot imagine what. At least not attributes that are set by the class itself. @rikigigi for the test that excepts, can you just print the content of the `results['output_trajectory'].attributes` just before it excepts. This apparently contains non-serializable data, but we need to understand why

rikigigi replied:
I added a `pdb.set_trace()` to inspect the structures. The offending arrays are located in output_parameters:
```
(aiida-venv) bertossa@pippo:~/sissa/aiida-quantumespresso/tests$ py.test -k 'test_cp_default'
============================================================================ test session starts ============================================================================
platform linux -- Python 3.6.9, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /home/bertossa/sissa/aiida-quantumespresso/tests, inifile: pytest.ini
plugins: regressions-1.0.6, datadir-1.3.1
collected 130 items / 125 deselected / 5 selected                                                                                                                           

parsers/test_cp.py .
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/bertossa/sissa/aiida-quantumespresso/tests/parsers/test_cp.py(46)test_cp_default()
-> data_regression.check({
(Pdb) pp results['output_parameters'].get_dict()
{'bands': {'bands': array([[[-12.07535278,  -6.28316664,  -4.02636205,  -3.17507374]]]),
           'bands_units': 'eV',
           'occupations': array([[[1., 1., 1., 1.]]])},
 'beta_real_space': False,
 'charge_density': './charge-density.dat',
 'constraint_mag': 0,
 'creator_name': 'cp',
 'creator_version': '6.5',
 'dft_exchange_correlation': None,
 'do_magnetization': False,
 'do_not_use_time_reversal': None,
 'exit_status': 0,
 'fermi_energy_units': 'eV',
 'fft_grid': [60, 60, 60],
 'format_name': 'QEXSD',
 'format_version': '19.03.04',
 'has_dipole_correction': False,
 'has_electric_field': False,
 'inversion_symmetry': False,
 'k_points': [[0.0, 0.0, 0.0]],
 'k_points_units': '1 / angstrom',
 'k_points_weights': [2.0],
 'last_successful_printout': 30,
 'lattice_symmetries': [],
 'lda_plus_u_calculation': False,
 'lkpoint_dir': False,
 'lsda': False,
 'magnetization_angle1': [0.0, 0.0],
 'magnetization_angle2': [0.0, 0.0],
 'no_time_rev_operations': None,
 'non_colinear_calculation': False,
 'number_of_atomic_wfc': 6,
 'number_of_bands': 4,
 'number_of_bravais_symmetries': None,
 'number_of_electrons': 8.0,
 'number_of_k_points': 1,
 'number_of_spin_components': 1,
 'number_of_symmetries': None,
 'q_real_space': False,
 'rho_cutoff': 4353.821352096,
 'rho_cutoff_units': 'eV',
 'smooth_fft_grid': [60, 60, 60],
 'spin_orbit_calculation': False,
 'spin_orbit_domag': False,
 'starting_magnetization': [0.0, 0.0],
 'structure': {'atomic_positions_units': 'Angstrom',
               'atoms': [['H',
                          [0.3077936327955275,
                           0.49074731600773003,
                           0.6020237422809962]],
                         ['H',
                          [-0.25285429616438865,
                           -0.2184308056415271,
                           -0.7702509858867927]],
                         ['O',
                          [-0.16627752338900367,
                           -0.26930737986161374,
                           0.2154517857488999]]],
               'cell': {'atoms': [['H',
                                   [0.3077936327955275,
                                    0.49074731600773003,
                                    0.6020237422809962]],
                                  ['H',
                                   [-0.25285429616438865,
                                    -0.2184308056415271,
                                    -0.7702509858867927]],
                                  ['O',
                                   [-0.16627752338900367,
                                    -0.26930737986161374,
                                    0.2154517857488999]]],
                        'lattice_vectors': [[5.2917720859, 0.0, 0.0],
                                            [0.0, 5.2917720859, 0.0],
                                            [0.0, 0.0, 5.2917720859]],
                        'volume': 148.1847095290449},
               'direct_lattice_vectors_units': 'Angstrom',
               'lattice_parameter': 5.2917720859,
               'lattice_parameter_xml': 10.0,
               'number_of_atoms': 3,
               'number_of_species': 2,
               'reciprocal_lattice_vectors': [[1.0, 0.0, 0.0],
                                              [0.0, 1.0, 0.0],
                                              [0.0, 0.0, 1.0]],
               'species': {'index': [1, 2],
                           'mass': [1.0, 16.0],
                           'pseudo': ['H_HSCV_PBE-1.0.upf',
                                      'O_HSCV_PBE-1.0.upf'],
                           'type': ['H', 'O']}},
 'symmetries': [],
 'symmetries_units': 'crystal',
 'time_reversal_flag': True,
 'wall_time': '     32.51s ',
 'warnings': [],
 'wfc_cutoff': -27.2113834506,
 'wfc_cutoff_units': 'eV'}
(Pdb) pp results['output_trajectory'].attributes
{'array|cell_temperature': [8],
 'array|cells': [8, 3, 3],
 'array|electronic_kinetic_energy': [8],
 'array|energy_constant_motion': [8],
 'array|enthalpy': [8],
 'array|enthalpy_plus_kinetic': [8],
 'array|forces': [0],
 'array|ionic_temperature': [8],
 'array|positions': [8, 3, 3],
 'array|pressure': [8],
 'array|scf_total_energy': [8],
 'array|steps': [8],
 'array|times': [8],
 'array|velocities': [8, 3, 3],
 'array|volume': [8],
 'symbols': ['H', 'H', 'O']}
(Pdb) 
```

rikigigi replied:
I added a new terrible fixture to solve the problem (see e2bbb7b)

rikigigi commented:
I

rikigigi commented:
I think it is ready for a second round of review! The code is less messy, and there are 4 more tests for the cp parser.

giovannipizzi commented on aiida_quantumespresso/calculations/__init__.py:
```diff
@@ -222,6 +221,7 @@ def prepare_for_submission(self, folder):
         calcinfo.retrieve_list.extend(self.xml_filepaths)
         calcinfo.retrieve_list += settings.pop('ADDITIONAL_RETRIEVE_LIST', [])
         calcinfo.retrieve_list += self._internal_retrieve_list
+        print(('calcinfo.retrieve_list = ', calcinfo.retrieve_list))
```
Can you remove this line?

giovannipizzi commented on aiida_quantumespresso/calculations/__init__.py:
```diff
@@ -240,6 +240,13 @@ def prepare_for_submission(self, folder):
+        # pylint: disable=unused-argument,invalid-name
+        """By default, nothing specific is generated.
+           This method can be implemented again in derived classes, and it will be called by _generate_PWCPinputdata"""
+        return u''
```
No need to specify `u` now that we only support python 3

giovannipizzi commented on aiida_quantumespresso/parsers/parse_raw/cp.py:
```diff
@@ -81,54 +88,57 @@ def parse_cp_text_output(data, xml_data):
+                    pass
+    except IndexError:
+        #when the cp does a cg, the output is different and the parser above does not work
+        pass
```
If you know it's because if `cg` maybe you can check this explicitly? Or maybe at least issue a log message that will end up in the report, or add a warning, just to know why we will not find some of the outputs?

giovannipizzi replied:
BTW - is this information also elsewhere? Can we avoid to parse the output file now?

giovannipizzi commented on aiida_quantumespresso/parsers/parse_raw/cp.py:
```diff
@@ -208,108 +240,120 @@ def parse_cp_xml_output(data):
     # CARD CONTROL
 
-    cardname = 'CONTROL'
+    cardname='CONTROL'
```
There are a bunch of stylistic changes here. We are using automatic styling: you probably want to install the pre-commit hooks and run them, to fix the problems.

Space problems are automatically fixed by yapf.

Then, there might be a few more suggestions by prospector that you will need to address to ensure the tests pass, see error log e.g. here: https://github.com/aiidateam/aiida-quantumespresso/pull/455/checks?check_run_id=732410782

giovannipizzi replied:
This guide from aiida-core should apply more or less also here: https://github.com/aiidateam/aiida-core/wiki/Development-environment#pre-commit-hooks

rikigigi commented:
Right now the tests on the cp parser are failing because of the upgrade of the library xmlschema: CP's xml is NOT compliant to the xml schema because many tags are missing, and the parsing fails. Is there any simple solution to tell the library to don't look at the schema and keep existing tags without rewriting all the parser? Why was it working before the upgrade of xmlschema?

giovannipizzi commented on aiida_quantumespresso/parsers/parse_xml/pw/parse.py:
```diff
@@ -50,26 +50,27 @@ def parse_xml(xml_file, dir_with_bands=None):
 
     # detect schema name+path from XML contents
-    schema_filepath = get_schema_filepath(xml)
+    schema_filepath = get_schema_filepath(xml, schema=schema)
```
This should be a more generic function, so ok to pass `schema='pw'` directly here. Also, the function should be moved, see comments below

rikigigi replied:
see 5fbc3ef5 major refactoring

giovannipizzi commented on aiida_quantumespresso/parsers/parse_xml/pw/parse.py:
```diff
@@ -167,10 +168,10 @@ def parse_pw_xml_post_6_2(xml):
     # See also PW/src/setup.f90
-    nsym = outputs['symmetries']['nsym']  # crystal symmetries
-    nrot = outputs['symmetries']['nrot']  # lattice symmetries
+    nsym = outputs.get('symmetries', {}).get('nsym', None)  # crystal symmetries
```
Is this optional only for CP?
BTW, if this function is reused by CP, we should move this function to a top-level folder (i.e. not leave it inside `parse_xml/pw/parse.py` but move it simply to `parse_xml/parse.py`).

rikigigi replied:
see 5fbc3ef5 major refactoring

giovannipizzi commented on aiida_quantumespresso/parsers/parse_xml/pw/parse.py:
```diff
@@ -252,20 +253,21 @@ def parse_pw_xml_post_6_2(xml):
         'number_of_bravais_symmetries': nrot,  # lattice symmetries
         'number_of_symmetries': nsym,  # crystal symmetries
-        'wfc_cutoff': inputs['basis']['ecutwfc'] * hartree_to_ev,
+        'wfc_cutoff': inputs.get('basis', {}).get('ecutwfc', -1.0) * hartree_to_ev,
```
Are there really cases where ecutwfc is not defined?

rikigigi replied:
This should be the issue with the still missing part of cp's xml file

giovannipizzi commented on aiida_quantumespresso/parsers/parse_xml/pw/schemas_cp/README.md:
```diff
@@ -0,0 +1,6 @@
+# Note on CP schema
```
With the new `xmlschema` library, do we still need two schemas folders? If so, let's not use the folder structure `parse_xml/pw/schemas_cp/README.md` (i.e. schemas-cp inside pw) but use `parse_xml/cp/schemas/`

rikigigi replied:
see 5fbc3ef5 major refactoring

giovannipizzi commented on aiida_quantumespresso/parsers/parse_xml/pw/versions.py:
```diff
@@ -30,14 +30,19 @@ def get_xml_file_version(xml):
 
 
-def get_schema_filepath(xml):
+def get_schema_filepath(xml, schema):
```
If we want to use this both for cp and pw, please move one level up in the folder structure (not inside `pw`). Otherwise, since the function is so simple you can just copy-paste the function twice for the pw and cp implementation, it's really just concatenating filenames.

rikigigi replied:
see 5fbc3ef5 major refactoring

giovannipizzi commented on setup.json:
```diff
@@ -88,7 +88,7 @@
         "packaging",
         "qe-tools~=1.1",
-        "xmlschema~=1.2"
+        "xmlschema>=1.2.5"
```
The correct syntax would be here `xmlschema~=1.2,>=1.2.5` (this allows also versions 1.3, 1.4, ..., but not go do 2.0 that might be backwards incompatible).

rikigigi replied:
modified in commit f2f78da3

giovannipizzi commented on tests/calculations/test_cp.py:
```diff
@@ -0,0 +1,42 @@
+
+    # Check the attributes of the returned `CalcInfo`
+    assert isinstance(calc_info, datastructures.CalcInfo)
+    assert sorted(calc_info.cmdline_params) == sorted(cmdline_params)
```
This was a bug in previous tests, this should become something like
`assert sorted(calc_info.codes[0].cmdline_params) == sorted(cmdline_params)`
(it's in the code_info)

rikigigi replied:
modified in commit f2f78da3

giovannipizzi commented on tests/calculations/test_cp.py:
```diff
@@ -0,0 +1,42 @@
+# -*- coding: utf-8 -*-
+"""Tests for the `PwCalculation` class."""
```
```suggestion
"""Tests for the `CpCalculation` class."""
```

giovannipizzi commented on tests/calculations/test_cp.py:
```diff
@@ -0,0 +1,42 @@
+def test_cp_autopilot(
+    aiida_profile, fixture_sandbox, generate_calc_job, generate_inputs_cp, file_regression, autopilot
+):
+    """Test a default `PwCalculation`."""
```
```suggestion
    """Test a default `CpCalculation`."""
```

giovannipizzi commented on tests/calculations/test_cp.py:
```diff
@@ -0,0 +1,42 @@
+import pytest
+
+
+@pytest.mark.parametrize('autopilot', [True, False])
+def test_cp_autopilot(
+    aiida_profile, fixture_sandbox, generate_calc_job, generate_inputs_cp, file_regression, autopilot
```
This is something changed recently
```suggestion
@pytest.mark.parametrize('autopilot', [True, False])
@pytest.mark.use_fixtures('aiida_profile')
def test_cp_autopilot(
    fixture_sandbox, generate_calc_job, generate_inputs_cp, file_regression, autopilot
```

giovannipizzi commented on tests/parsers/test_cp.py:
```diff
@@ -1,38 +1,62 @@
-    })
+    if version != '6.5_cgstep':
+        assert 'output_trajectory' in results
+        if version == '6.5_autopilot':
+            pass
+            #pdb.set_trace()
```
Useless, to be removed

rikigigi replied:
modified in commit f2f78da3

giovannipizzi commented on tests/parsers/fixtures/cp/retrieve.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
```
What is this file for? Maybe it's not needed to commit

giovannipizzi commented on tests/parsers/test_cp.py:
```diff
@@ -1,38 +1,62 @@
 # -*- coding: utf-8 -*-
 # pylint: disable=redefined-outer-name
 """Tests for the `CpParser`."""
+#import pdb
```
Please remove

rikigigi replied:
modified in commit f2f78da3

giovannipizzi commented on tests/parsers/test_cp.py:
```diff
@@ -1,38 +1,62 @@
+            #pdb.set_trace()
+        data_regression.check({
+            'parameters':
+            call_something(numpy.ndarray, 'tolist', results['output_parameters'].get_dict(), func=call_something),
```
Are you sure the `output_parameters` dict contains numpy arrays? This should be stored as a JSON dictionary so just plain lists/dicts. Maybe you wanted to do it for the ArrayData nodes?

rikigigi replied:
Maybe there was an issue in the code somewhere else, it was a lot of time ago... I'll check in the next days

rikigigi replied:
I confirm that output_parameters contains numpy arrays

giovannipizzi commented on tests/parsers/test_cp.py:
```diff
@@ -1,38 +1,62 @@
     assert calcfunction.is_finished, calcfunction.exception
     assert calcfunction.is_finished_ok, calcfunction.exit_message
-    assert not orm.Log.objects.get_logs_for(node)
+    if version != '6.5_cgstep':
```
From this line below, we suggest to simplify the logic as follows:
```python
assert 'output_parameters' in results 
data = {'parameters': results['output_parameters'],}

if version != '6.5_cgstep': 
    assert not orm.Log.objects.get_logs_for(node)
    assert 'output_trajectory' in results 
    data['output_trajectory'] = results['output_trajectory']

data_regression(data) 
```

rikigigi replied:
modified in commit f2f78da3

giovannipizzi reviewed (changes requested):
Hi @rikigigi thanks a lot for all the work!
@sphuber and me went carefully through the PR.
We left some easy comments to fix/simplify the code, and a couple of questions.

If you could address them, and rebase your work on the current develop branch (so removing all conflicts) we think that then this is ready to be merged!

Let us know if you have issues when merging.

rikigigi commented:
> Hi @rikigigi thanks a lot for all the work!
> @sphuber and me went carefully through the PR.
> We left some easy comments to fix/simplify the code, and a couple of questions.
> 
> If you could address them, and rebase your work on the current develop branch (so removing all conflicts) we think that then this is ready to be merged!
> 
> Let us know if you have issues when merging.

Thank you very much, I'll write to you in the next few days when I'm done

rikigigi commented on aiida_quantumespresso/parsers/parse_raw/cp.py:
```diff
@@ -31,7 +38,10 @@ def parse_cp_traj_stanzas(num_elements, splitlines, prepend_name, rescale=1.):
             if len(l) == 2:
                 steps.append(int(l[0]))
-                times.append(float(l[1]))
+                if l[1] == '***********':
```
i fixed it in QE some weeks ago. I don't know, the code was printing exactly this number of `*` because of the fortran format...

rikigigi commented on aiida_quantumespresso/calculations/cp.py:
```diff
@@ -153,3 +160,36 @@ def define(cls, spec):
+        #input_params = kwargs['input_params']
+        settings = kwargs['settings']
+        #AUTOPILOT
+        inputfile = u''
```
ok: ffdcdcb410

rikigigi commented on aiida_quantumespresso/parsers/parse_xml/pw/parse.py:
```diff
@@ -50,26 +50,27 @@ def parse_xml(xml_file, dir_with_bands=None):
 
     if xml_file_version == QeXmlVersion.POST_6_2:
-        parsed_data, logs = parse_pw_xml_post_6_2(xml_parsed)
+        parsed_data, logs = parse_xml_post_6_2(xml_parsed, schema='pw')
```
see 5fbc3ef5 major refactoring

rikigigi commented:
I have to check what is wrong with the merge, since some tests not related with my work are failing :( .
By the way I think that this pull request is becoming really too huge, keeping it updated and out of tree is becoming a demanding task

rikigigi commented:
The issues that I found with the merge are fixed. For me ready to merge

sphuber commented:
@rikigigi I am working on going through the code. Not sure if I can finish it today, but otherwise I will continue soon next week. Thanks

sphuber reviewed (commented):
@rikigigi I squashed your changes into a single commit and then added one more commit on top with changes that I would propose. Please give them a look to see if you see anything out of the ordinary. It is mostly cleaning up some logic and removing some stale code you added that is no longer needed.

There is one significant change. As mentioned in one of the comments in the review, I removed the `bands` and `structure` keys from the `output_parameters`. As I understand it, these are parsed from the new XML format and are also what caused you to require to convert these numpy arrays to lists in the parser test. I personally don't think the structure and bands should be stored in the `output_parameters`. If they are important to keep, they should become their own output node, just like for `PwCalculation`. Apparently those outputs don't exist yet for `CpCalculation` so I guess they maybe don't make sense. @giovannipizzi do you agree to remove these keys from the parsed data?

Finally, please check the comment on the output test files coming from an unreleased QE version. @giovannipizzi what do you think on that point?

rikigigi commented on tests/parsers/test_cp/test_cp_default_6_5_autopilot_.yml:
```diff
@@ -0,0 +1,188 @@
+      - - 0.0
+        - 0.0
+        - 5.2917720859
+      volume: 148.1847095290449
```
This was always the case, I don't think that I touched that part of the plugin! The only new feature that I coded was the autopilot input.

sphuber replied:
I think it is just because of the new XML. The old XML did not include the structure and bands, or at least the old parser did not parse them and put them in the `output_parameters`. We can see this from the existing `default` parser test, where the `output_parameters` do not include these quantities. Anyway, we had the same problem when we first added support for the new XML for the `PwParser` and we spend a great deal of time to make sure that the keys in `output_parameters` would remain the same. This is why we added this big mapping function in `parse_xml/pw/parse.py` where the data parsed automatically from the XML using the schema library, is mapped onto different keys that were also produced by the legacy parser. @giovannipizzi this is clearly not done here. Fine to keep it as is, accepting that the structure of `output_parameters` will depend heavily on the version of the XML output file?

giovannipizzi replied:
Hi, it's a good point - ideally I would suggest to do what we did in pw, but in practice this is probably just too much work and I don't know if we have anyone who would have the time to do it; in the end, I think it's better to have a plugin that is usable even if the content of the output dictionary has changed (at the moment I think it's not working with the new XML). 

The only thing that I don't know if it's worth trying to copy-paste the mapping function, trying to apply to CP and do the minimal modifications to make it work for CP. But if it's too complex, better to merge this.

Regarding bands and structure - I think it's OK to remove them, especially if they weren't there before.
For the structure, this should be in general taken from the last snapshot of the trajectory.
Not sure about the bands, but these should be probably in one of the output arrays? In any case, I agree that if we really want to keep them we should put them in a BandsData.
And if the information is in the XML that is anyway stored in AiiDA, these can always be parsed later, and we can extend later the parser to also output those in a backward-compatible way.

rikigigi replied:
Isn't at this moment the xml parser doing exactly the same both for cp and pw? If I understood correctly what you are saying, the dictionary that you are talking about is https://github.com/rikigigi/aiida-quantumespresso/blob/baeec661bf03a4b7c53c3ae34f4ff7caad62d350/aiida_quantumespresso/parsers/parse_xml/parse.py#L200-L253 , and this code is called by cp's parser also. Regarding the legacy cp and pw parser, I didn't touched them

rikigigi commented on tests/parsers/test_cp.py:
```diff
@@ -6,31 +6,53 @@
 
-
-def test_cp_default(fixture_localhost, generate_calc_job_node, generate_parser, generate_inputs, data_regression):
+#Now I use a post-6.5 git version (order of atom in the output changed)
```
This was a long time ago. Two months ago the 6.6 version was released... I hope that the output will not change anymore, but the xml (that is still incomplete and there is some work in progress).

giovannipizzi replied:
So @rikigigi would it make sense to have the tests for the released 6.6?
And for the fix on the short format, you mean the change in the parsed that sets a default value when it finds `****`? Or a fix in QE?

rikigigi replied:
Yes, it would make sense, the output should be exactly the same! The fix for the short format is in both the parser and in QE

rikigigi replied:
I prepared a script for generating the outputs and I added the test for version 6.6

rikigigi commented:
> @rikigigi I squashed your changes into a single commit and then added one more commit on top with changes that I would propose. Please give them a look to see if you see anything out of the ordinary. It is mostly cleaning up some logic and removing some stale code you added that is no longer needed.
> 
> There is one significant change. As mentioned in one of the comments in the review, I removed the `bands` and `structure` keys from the `output_parameters`. As I understand it, these are parsed from the new XML format and are also what caused you to require to convert these numpy arrays to lists in the parser test. I personally don't think the structure and bands should be stored in the `output_parameters`. If they are important to keep, they should become their own output node, just like for `PwCalculation`. Apparently those outputs don't exist yet for `CpCalculation` so I guess they maybe don't make sense. @giovannipizzi do you agree to remove these keys from the parsed data?
> 
> Finally, please check the comment on the output test files coming from an unreleased QE version. @giovannipizzi what do you think on that point?

For me the changes are ok!

rikigigi commented:
@sphuber after your modifications some test are failing... is it ok?  shall I do --force-regen on all the tests?
There are again conflicts...

sphuber commented:
Thanks @rikigigi I will have a look at the failing test and rebase to address the conflicts. Regarding the new output files, is there any significant difference between the outputs of 6.5 and 6.6? If not, I would just keep that of 6.6 in order to not include too many unnecessary files in the repo

rikigigi commented:
yes, they were basically the same, I removed the non official 6.5 git output

sphuber commented:
OK, I finally found the source of the failing tests, there was a very subtle bug in the code that determines the filepath of the XML schema, and due to it, it was always silently falling back onto the default, which is outdated and didn't parse all the outputs. I will now merge this. Thanks a lot for the work and patience @rikigigi  apologies for the wait.
